### PR TITLE
wip research ability to create simple report of done tasks

### PR DESCRIPTION
**Pull request by titoOdUA**
I have started exploring of the ability to create simple reports discussed here https://github.com/alanvardy/tod/issues/355#issuecomment-1567364369

### CODE REVIEW

**Comment by titoOdUA** [comment]
it is just a rough draft, I don't have much time for now. I will have time on Friday/Saturday.

@alanvardy how do you like `completed_at` field in tasks? it is in UTC by default and private, we can get it through getter which will convert it to the user's timezone.

Also, I see that we have very similar functions in the task list 
```rust
    if has_flag(matches.clone(), "scheduled") {
        projects::scheduled_items(&config, &project)
    } else if has_flag(matches.clone(), "done-yesterday") {
        projects::completed_items(&config, &project)
    } else {
        projects::all_items(&config, &project)
    }

```
scheduled_items, completed_items, all_items
I think that we can extract all of it in a separate file/module that would have some enum of available variants. We can have one functions that make requests based on the provided variant of the enum. And maybe it could receive some closure as a parameter for filtering purposes. (maybe our task list can transform into some module for reporting/filtering)

@alanvardy how do you like the idea?

**Comment by alanvardy** [comment]
> it is just a rough draft, I don't have much time for now. I will have time on Friday/Saturday.

:+1: 

> @alanvardy how do you like `completed_at` field in tasks? it is in UTC by default and private, we can get it through getter which will convert it to the user's timezone.

I like it!

> Also, I see that we have very similar functions in the task list
> 
> ```rust
>     if has_flag(matches.clone(), "scheduled") {
>         projects::scheduled_items(&config, &project)
>     } else if has_flag(matches.clone(), "done-yesterday") {
>         projects::completed_items(&config, &project)
>     } else {
>         projects::all_items(&config, &project)
>     }
> ```
> 
> scheduled_items, completed_items, all_items I think that we can extract all of it in a separate file/module that would have some enum of available variants. We can have one functions that make requests based on the provided variant of the enum. And maybe it could receive some closure as a parameter for filtering purposes. (maybe our task list can transform into some module for reporting/filtering)
> 
> @alanvardy how do you like the idea?

Yes there is lots of duplicate code in there, I'm interested in seeing what you come up with!

File: src/items.rs [hunk 1]
```diff
@@ -9,6 +10,7 @@ use std::cmp::Reverse;
 use std::fmt::Display;
 
 use crate::config::Config;
+use crate::time::timezone_from_str;
```
Comment (alanvardy) [inline src/items.rs:13]:
```suggestion
use crate::time;
```

File: src/items.rs [hunk 2]
```diff
@@ -253,6 +256,13 @@ impl Item {
             Ok(DateTimeInfo::DateTime { .. })
         )
     }
+
+    pub fn get_completed_at(&self, config: &Config) -> DateTime<Tz> {
+        let tz = timezone_from_str(&config.timezone);
```
Comment (alanvardy) [inline src/items.rs:261]:
```suggestion
        let tz = time::timezone_from_str(&config.timezone);
```

**Review by alanvardy — COMMENTED** [review]
So far so good, I'm interested in using this feature when it is done

**Comment by alanvardy** [comment]
Sorry about that, I'm hoping that by splitting out into more modules we can conflict less in the future. There's a buncha spaghetti code in there!

**Pull request closed without merging**